File: src/deep-analysis.js
const path = require('path');
require('dotenv').config({ path: path.join(__dirname, '../config/.env') });
const { createClient } = require('@supabase/supabase-js');

const supabaseUrl = process.env.SUPABASE_URL;
const supabaseKey = process.env.SUPABASE_SERVICE_KEY || process.env.SUPABASE_ANON_KEY || process.env.SUPABASE_KEY;

const supabase = createClient(supabaseUrl, supabaseKey);

async function deepAnalysis() {
  const experimentId = '8f688916-a7a7-4501-badc-6cc3a5efc8d8';

  console.log('='.repeat(80));
  console.log('实验深度分析报告');
  console.log('实验ID:', experimentId);
  console.log('='.repeat(80));

  // 获取所有数据
  const { data: tokens } = await supabase
    .from('experiment_tokens')
    .select('*')
    .eq('experiment_id', experimentId);

  const { data: signals } = await supabase
    .from('strategy_signals')
    .select('*')
    .eq('experiment_id', experimentId)
    .order('created_at', { ascending: true });

  // 按代币分组分析
  const tokenAnalysis = {};

  // 初始化代币分析对象
  tokens.forEach(t => {
    tokenAnalysis[t.token_address] = {
      symbol: t.token_symbol,
      address: t.token_address,
      status: t.status,
      discoveredAt: new Date(t.discovered_at),
      buySignals: [],
      sellSignals: [],
      trades: [],
      apiData: t.raw_api_data ? (typeof t.raw_api_data === 'string' ? JSON.parse(t.raw_api_data) : t.raw_api_data) : null
    };
  });

  // 分析信号
  signals.forEach(s => {
    const addr = s.token_address;
    if (tokenAnalysis[addr]) {
      if (s.signal_type === 'BUY') {
        tokenAnalysis[addr].buySignals.push({
          createdAt: new Date(s.created_at),
          reason: s.reason,
          metadata: s.metadata ? (typeof s.metadata === 'string' ? JSON.parse(s.metadata) : s.metadata) : null
        });
      } else if (s.signal_type === 'SELL') {
        tokenAnalysis[addr].sellSignals.push({
          createdAt: new Date(s.created_at),
          reason: s.reason,
          metadata: s.metadata ? (typeof s.metadata === 'string' ? JSON.parse(s.metadata) : s.metadata) : null
        });
      }
    }
  });

  // 计算每个代币的盈亏
  console.log('\n【已买入代币盈亏分析】');
  console.log('='.repeat(80));

  const boughtTokens = Object.values(tokenAnalysis).filter(t => t.status === 'bought');
  let totalInvested = 0;
  let totalReturnValue = 0;
  let totalProfit = 0;
  const profitLossList = [];

  boughtTokens.forEach(t => {
    if (t.buySignals.length > 0 && t.sellSignals.length > 0) {
      const buySignal = t.buySignals[0];
      const buyPrice = buySignal.metadata?.price || buySignal.metadata?.tradeResult?.trade?.unit_price || 0;

      // 找到最后一次卖出信号
      const lastSellSignal = t.sellSignals[t.sellSignals.length - 1];
      const sellPrice = lastSellSignal.metadata?.price || 0;
      const profitPercent = lastSellSignal.metadata?.profitPercent || 0;

      const invested = buySignal.metadata?.cardConfig?.totalCards * buySignal.metadata?.cardConfig?.perCardMaxBNB || 0.4;
      const returnValue = invested * (1 + profitPercent / 100);
      const profit = returnValue - invested;

      totalInvested += invested;
      totalReturnValue += returnValue;
      totalProfit += profit;

      profitLossList.push({
        symbol: t.symbol,
        address: t.address,
        buyPrice,
        sellPrice,
        profitPercent,
        invested,
        returnValue,
        profit,
        holdDuration: (lastSellSignal.metadata?.holdDuration / 60).toFixed(2) + '分钟'
      });
    }
  });

  // 按盈亏排序
  profitLossList.sort((a, b) => b.profitPercent - a.profitPercent);

  console.log('\n代币盈亏明细（按收益率排序）：');
  console.log('代币\t\t买入价\t\t卖出价\t\t收益率\t\t投入\t\t回报\t\t利润\t\t持仓时间');
  console.log('-'.repeat(120));
  profitLossList.forEach(p => {
    console.log(`${p.symbol}\t\t${p.buyPrice}\t\t${p.sellPrice}\t\t${p.profitPercent.toFixed(2)}%\t\t${p.invested}\t\t${p.returnValue.toFixed(2)}\t\t${p.profit.toFixed(2)}\t\t${p.holdDuration}`);
  });

  console.log('\n【总体盈亏统计】');
  console.log('总投入:', totalInvested.toFixed(2), 'BNB');
  console.log('总回报:', totalReturnValue.toFixed(2), 'BNB');
  console.log('总盈亏:', totalProfit.toFixed(2), 'BNB');
  console.log('总收益率:', (totalProfit / totalInvested * 100).toFixed(2), '%');

  // 分析盈利和亏损代币
  const profitable = profitLossList.filter(p => p.profitPercent > 0);
  const lossMaking = profitLossList.filter(p => p.profitPercent < 0);

  console.log('\n【盈利代币分析】');
  console.log('盈利代币数:', profitable.length);
  if (profitable.length > 0) {
    const avgProfitPercent = profitable.reduce((sum, p) => sum + p.profitPercent, 0) / profitable.length;
    const totalProfit = profitable.reduce((sum, p) => sum + p.profit, 0);
    console.log('平均收益率:', avgProfitPercent.toFixed(2), '%');
    console.log('总利润:', totalProfit.toFixed(2), 'BNB');
    console.log('\n表现最好的5个代币：');
    profitable.slice(0, 5).forEach(p => {
      console.log(`  ${p.symbol}: ${p.profitPercent.toFixed(2)}% (利润: ${p.profit.toFixed(2)} BNB)`);
    });
  }

  console.log('\n【亏损代币分析】');
  console.log('亏损代币数:', lossMaking.length);
  if (lossMaking.length > 0) {
    const avgLossPercent = lossMaking.reduce((sum, p) => sum + p.profitPercent, 0) / lossMaking.length;
    const totalLoss = lossMaking.reduce((sum, p) => sum + p.profit, 0);
    console.log('平均亏损率:', avgLossPercent.toFixed(2), '%');
    console.log('总亏损:', totalLoss.toFixed(2), 'BNB');
    console.log('\n亏损最严重的5个代币：');
    lossMaking.reverse().slice(0, 5).forEach(p => {
      console.log(`  ${p.symbol}: ${p.profitPercent.toFixed(2)}% (亏损: ${p.profit.toFixed(2)} BNB)`);
    });
  }

  // 分析止损策略
  console.log('\n【止损策略分析】');
  const stopLossTrades = profitLossList.filter(p => p.profitPercent < -30);
  console.log('触发止损的代币数（亏损>30%）:', stopLossTrades.length);
  if (stopLossTrades.length > 0) {
    stopLossTrades.forEach(p => {
      console.log(`  ${p.symbol}: ${p.profitPercent.toFixed(2)}%`);
    });
  }

  // 分析监控中可能错失的机会
  console.log('\n【监控中代币特征分析】');
  const monitoringTokens = Object.values(tokenAnalysis).filter(t => t.status === 'monitoring');

  // 按发现时间分析，看看有没有错过的机会
  const earlyTokens = monitoringTokens.filter(t => {
    const age = (Date.now() - new Date(t.discoveredAt).getTime()) / 1000 / 60; // 分钟
    return age > 30; // 超过30分钟的代币
  });

  console.log('监控中超过30分钟的代币数:', earlyTokens.length);
  console.log('这些代币可能已被系统过滤，分析其特征：');

  // 定义统计变量
  let fdvStats = {
    min: Infinity,
    max: 0,
    avg: 0,
    hasData: false
  };
  let tvlStats = {
    min: Infinity,
    max: 0,
    avg: 0,
    hasData: false
  };

  if (earlyTokens.length > 0) {
    // 分析这些代币的FDV、TVL等特征

    earlyTokens.forEach(t => {
      if (t.apiData) {
        const fdv = parseFloat(t.apiData.fdv) || 0;
        const tvl = parseFloat(t.apiData.tvl) || 0;
        if (fdv > 0) {
          fdvStats.min = Math.min(fdvStats.min, fdv);
          fdvStats.max = Math.max(fdvStats.max, fdv);
          fdvStats.avg += fdv;
          fdvStats.hasData = true;
        }
        if (tvl > 0) {
          tvlStats.min = Math.min(tvlStats.min, tvl);
          tvlStats.max = Math.max(tvlStats.max, tvl);
          tvlStats.avg += tvl;
          tvlStats.hasData = true;
        }
      }
    });

    fdvStats.avg /= earlyTokens.filter(t => t.apiData?.fdv).length || 1;
    tvlStats.avg /= earlyTokens.filter(t => t.apiData?.tvl).length || 1;

    console.log('\nFDV范围:', fdvStats.min === Infinity ? 'N/A' : fdvStats.min.toFixed(2), '-', fdvStats.max.toFixed(2));
    console.log('FDV平均:', fdvStats.avg.toFixed(2));
    console.log('TVL范围:', tvlStats.min === Infinity ? 'N/A' : tvlStats.min.toFixed(2), '-', tvlStats.max.toFixed(2));
    console.log('TVL平均:', tvlStats.avg.toFixed(2));

    // 显示一些示例
    console.log('\n示例代币：');
    earlyTokens.slice(0, 10).forEach(t => {
      console.log(`  ${t.symbol || 'N/A'}: FDV=${t.apiData?.fdv || 'N/A'}, TVL=${t.apiData?.tvl || 'N/A'}`);
    });
  }

  // 分析已买入代币的特征
  console.log('\n【已买入代币特征分析】');
  const boughtTokensData = Object.values(tokenAnalysis).filter(t => t.status === 'bought');

  const boughtFdvStats = {
    min: Infinity,
    max: 0,
    avg: 0
  };
  const boughtTvlStats = {
    min: Infinity,
    max: 0,
    avg: 0
  };

  boughtTokensData.forEach(t => {
    if (t.apiData) {
      const fdv = parseFloat(t.apiData.fdv) || 0;
      const tvl = parseFloat(t.apiData.tvl) || 0;
      if (fdv > 0) {
        boughtFdvStats.min = Math.min(boughtFdvStats.min, fdv);
        boughtFdvStats.max = Math.max(boughtFdvStats.max, fdv);
        boughtFdvStats.avg += fdv;
      }
      if (tvl > 0) {
        boughtTvlStats.min = Math.min(boughtTvlStats.min, tvl);
        boughtTvlStats.max = Math.max(boughtTvlStats.max, tvl);
        boughtTvlStats.avg += tvl;
      }
    }
  });

  boughtFdvStats.avg /= boughtTokensData.filter(t => t.apiData?.fdv).length || 1;
  boughtTvlStats.avg /= boughtTokensData.filter(t => t.apiData?.tvl).length || 1;

  console.log('FDV范围:', boughtFdvStats.min === Infinity ? 'N/A' : boughtFdvStats.min.toFixed(2), '-', boughtFdvStats.max.toFixed(2));
  console.log('FDV平均:', boughtFdvStats.avg.toFixed(2));
  console.log('TVL范围:', boughtTvlStats.min === Infinity ? 'N/A' : boughtTvlStats.min.toFixed(2), '-', boughtTvlStats.max.toFixed(2));
  console.log('TVL平均:', boughtTvlStats.avg.toFixed(2));

  // 对比分析
  console.log('\n【特征对比分析】');
  console.log('已买入代币 vs 监控中代币的特征差异：');
  if (fdvStats.hasData && tvlStats.hasData) {
    console.log('FDV差异:', boughtFdvStats.avg.toFixed(2), '(买入) vs', fdvStats.avg.toFixed(2), '(监控)');
    console.log('TVL差异:', boughtTvlStats.avg.toFixed(2), '(买入) vs', tvlStats.avg.toFixed(2), '(监控)');
  } else {
    console.log('监控中代币无有效特征数据');
  }

  // 分析卖出策略触发情况
  console.log('\n【卖出策略触发统计】');
  const sellReasonCount = {};
  signals.filter(s => s.signal_type === 'SELL').forEach(s => {
    const reason = s.reason;
    sellReasonCount[reason] = (sellReasonCount[reason] || 0) + 1;
  });
  Object.entries(sellReasonCount).forEach(([reason, count]) => {
    console.log(`  ${reason}: ${count} 次`);
  });

  console.log('\n='.repeat(80));
  console.log('分析完成');
  console.log('='.repeat(80));
}

deepAnalysis().catch(console.error);

File: deep_analysis.py
#!/usr/bin/env python3
"""
深入分析实验 8f688916-a7a7-4501-badc-6cc3a5efc8d8
重点分析: 为什么买入的代币都亏损
"""

import json
import statistics
from datetime import datetime

# 加载数据
with open('/tmp/tokens.json') as f:
    tokens_data = json.load(f)

with open('/tmp/trades.json') as f:
    trades_data = json.load(f)

# 处理代币数据，建立映射
token_map = {}
for token in tokens_data.get('data', []):
    addr = token.get('token_address')
    raw = token.get('raw_api_data', {})

    # 解析TVL和FDV
    try:
        tvl = float(raw.get('tvl', 0)) if raw.get('tvl') not in ['0', '', None] else 0
    except:
        tvl = 0

    try:
        fdv = float(raw.get('fdv', 0)) if raw.get('fdv') not in ['0', '', None] else 0
    except:
        fdv = 0

    token_map[addr] = {
        'symbol': token.get('token_symbol'),
        'tvl': tvl,
        'fdv': fdv,
        'platform': raw.get('issue_platform', ''),
        'launch_at': raw.get('launch_at', 0),
        'discovered_at': token.get('discovered_at'),
        'status': token.get('status')
    }

# 分析交易
trades = trades_data.get('data', [])
buy_trades = [t for t in trades if t.get('trade_direction') == 'buy']
sell_trades = [t for t in trades if t.get('trade_direction') == 'sell']

print("=" * 100)
print(" " * 35 + "实验深度分析报告")
print("=" * 100)

# 1. 总体统计
print("\n【一、总体统计】")
print(f"总交易数: {len(trades)}")
print(f"买入交易: {len(buy_trades)}")
print(f"卖出交易: {len(sell_trades)}")

# 2. 按代币分析盈亏
print("\n【二、每笔交易详情】")
print(f"{'代币':<15} {'买入价格':<18} {'卖出价格':<18} {'盈亏%':<10} {'持仓秒':<12} {'TVL':<12} {'FDV':<12}")
print("-" * 110)

trade_results = []

for buy in buy_trades:
    token_addr = buy.get('token_address')
    symbol = buy.get('token_symbol')
    buy_price = buy.get('unit_price')
    buy_time = datetime.fromisoformat(buy.get('executed_at', buy.get('created_at')).replace('Z', '+00:00'))

    # 找到对应的卖出
    for sell in sell_trades:
        sell_time = datetime.fromisoformat(sell.get('executed_at', sell.get('created_at')).replace('Z', '+00:00'))
        if sell.get('token_address') == token_addr and sell_time > buy_time:
            sell_price = sell.get('unit_price')
            metadata = sell.get('metadata', {})
            profit_pct = metadata.get('profitPercent', 0)
            hold_duration = metadata.get('holdDuration', 0)

            # 从token_map获取TVL/FDV
            token_info = token_map.get(token_addr, {})
            tvl = token_info.get('tvl', 0)
            fdv = token_info.get('fdv', 0)

            trade_results.append({
                'symbol': symbol,
                'profit': profit_pct,
                'hold_seconds': hold_duration,
                'tvl': tvl,
                'fdv': fdv,
                'buy_price': buy_price,
                'sell_price': sell_price
            })

            print(f"{symbol:<15} {buy_price:<18.2e} {sell_price:<18.2e} {profit_pct:<10.2f} {hold_duration:<12.0f} ${tvl:<11,.2f} ${fdv:<11,.2f}")
            break

# 3. 盈亏统计
profits = [t['profit'] for t in trade_results]
losses = [p for p in profits if p < 0]
gains = [p for p in profits if p > 0]

print("\n【三、盈亏统计】")
print(f"总交易笔数: {len(trade_results)}")
print(f"盈利笔数: {len(gains)}")
print(f"亏损笔数: {len(losses)}")
print(f"平均盈亏: {statistics.mean(profits):.2f}%")
print(f"最大盈利: {max(profits):.2f}%")
print(f"最大亏损: {min(profits):.2f}%")
print(f"盈亏比: {len(gains)/len(losses) if losses else 0:.2f}")

# 4. 持仓时间分析
hold_times = [t['hold_seconds'] for t in trade_results]
print("\n【四、持仓时间分析】")
print(f"平均持仓: {statistics.mean(hold_times):.0f} 秒 ({statistics.mean(hold_times)/60:.2f} 分钟)")
print(f"最短持仓: {min(hold_times):.0f} 秒")
print(f"最长持仓: {max(hold_times):.0f} 秒 ({max(hold_times)/3600:.2f} 小时)")

# 5. TVL分析
tvls = [t['tvl'] for t in trade_results if t['tvl'] > 0]
print("\n【五、买入代币TVL分析】")
print(f"有TVL数据的代币: {len(tvls)} 个")
if tvls:
    print(f"TVL范围: ${min(tvls):.2f} - ${max(tvls):.2f}")
    print(f"平均TVL: ${statistics.mean(tvls):.2f}")
    print(f"中位数TVL: ${statistics.median(tvls):.2f}")

# 关键发现：所有已交易代币的TVL都是0！
zero_tvl_trades = [t for t in trade_results if t['tvl'] == 0]
print(f"\n⚠️  关键发现: {len(zero_tvl_trades)} 笔交易买入时TVL为0 (流动性极低!)")

# 6. 监控中代币的TVL分布
all_tokens_tvls = [t['tvl'] for t in token_map.values() if t['tvl'] > 0]
print("\n【六、监控中代币TVL分布】")
print(f"总监控代币: {len(token_map)}")
print(f"有TVL数据的: {len(all_tokens_tvls)}")
if all_tokens_tvls:
    all_tokens_tvls_sorted = sorted(all_tokens_tvls)
    print(f"TVL中位数: ${statistics.median(all_tokens_tvls):.2f}")
    print(f"TVL 25分位: ${all_tokens_tvls_sorted[len(all_tokens_tvls_sorted)//4]:.2f}")
    print(f"TVL 75分位: ${all_tokens_tvls_sorted[len(all_tokens_tvls_sorted)*3//4]:.2f}")

# 7. 亏损和盈利交易的持仓时间对比
loss_holds = [t['hold_seconds']/60 for t in trade_results if t['profit'] < 0]
gain_holds = [t['hold_seconds']/60 for t in trade_results if t['profit'] > 0]

print("\n【七、持仓时间对比】")
print(f"亏损交易平均持仓: {statistics.mean(loss_holds):.2f} 分钟")
print(f"盈利交易平均持仓: {statistics.mean(gain_holds):.2f} 分钟")

# 8. 价格变化分析
print("\n【八、价格变化分析】")
price_changes = []
for t in trade_results:
    if t['buy_price'] > 0:
        change = (t['sell_price'] - t['buy_price']) / t['buy_price'] * 100
        price_changes.append(change)

if price_changes:
    print(f"价格下跌平均: {statistics.mean([c for c in price_changes if c < 0]):.2f}%")
    print(f"价格上涨平均: {statistics.mean([c for c in price_changes if c > 0]):.2f}%")

print("\n" + "=" * 100)
